### PR TITLE
docs(idiomatic maya): add best practices guide to writing code in idiomatic maya based patterns

### PR DESCRIPTION
**Pull request by sagarkrsd**
**What this PR does / why we need it**:
This PR updates the idiomatic maya guide as per the standards being followed for making maya code idiomatic

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

Signed-off-by: sagarkrsd <sagar.kumar@openebs.io>

### CODE REVIEW

File: docs/idiomatic-maya-guide.md [hunk 1]
```diff
@@ -1,99 +1,179 @@
...
+// Good
+// upgraderesult looks more specific and to the point i.e. the alias
+// is for 'result' which is inside upgrade
+upgraderesult "github.com/openebs/maya/pkg/upgrade/result/v1alpha1"
+```
+
+```go
+// Bad
+// taskPatch represents details for runtask patch operation
+type taskPatch struct {}
```
Comment (AmitKumarDas) [inline docs/idiomatic-maya-guide.md]: This naming depends on package name as well.
So `patch` can be valid as well.

File: docs/idiomatic-maya-guide.md [hunk 2]
```diff
@@ -1,99 +1,179 @@
...
+// refer - https://blog.golang.org/package-names
+// refer the other naming idioms mentioned in this doc
+
+// Not Ok
+pkg/upgrade/nodeSelect/v1alpha1
+
+// vs.
+
+// Good (no camelcase)
+pkg/nodeselect/v1alpha1
```
Comment (AmitKumarDas) [inline docs/idiomatic-maya-guide.md]: `pkg/upgrade/nodeselect/v1alpha1`

File: docs/idiomatic-maya-guide.md [hunk 3]
```diff
@@ -1,99 +1,179 @@
...
+        return nil, err
+    }
+    return k.get(cs, name, k.namespace, opts)
 }
+// Here the caller code will import this package as
+upgraderesult "github.com/openebs/maya/pkg/upgrade/result/v1alpha1"
 
-vs.
+// And then the above method will be called as
+upgraderesult.GetUpgradeResult(name,opts)
```
Comment (AmitKumarDas) [inline docs/idiomatic-maya-guide.md]: `upgraderesult.KubeClient().GetUpgradeResult(name, opts)`

File: docs/idiomatic-maya-guide.md [hunk 4]
```diff
@@ -1,99 +1,179 @@
...
+    if err != nil {
+        return nil, err
+    }
+    return k.get(cs, name, k.namespace, opts)
+}
+// Here the caller code will import this package as
+upgraderesult "github.com/openebs/maya/pkg/upgrade/result/v1alpha1"
+
+// And then the above method will be called as
+upgraderesult.Get(name,opts)
```
Comment (AmitKumarDas) [inline docs/idiomatic-maya-guide.md]: `upgraderesult.KubeClient().Get(name, opts)`

File: docs/idiomatic-maya-guide.md [hunk 5]
```diff
@@ -1,99 +1,179 @@
...
+p, err := patch.
+  BuilderForRuntask("UpgradeResult", m.runtask.Spec.Task, m.templateValues).
+  AddCheck(patch.IsValidType()).
+  Build()
 ```
 
-### Do the names reflect their purpose?
 ```go
-// assume a file pkg/algorithm/nodeSelect/v1alpha1/select_node.go 
+// Bad
```
Comment (AmitKumarDas) [inline docs/idiomatic-maya-guide.md]: We might want to have separate go based snippet for bad vs. good code.
e.g.
```go
// bad code
```
```go
// good code
```

Notice two code snippets & not a single go snippet.

File: docs/idiomatic-maya-guide.md (hunk 1 above)
Comment (sagarkrsd) [reply docs/idiomatic-maya-guide.md]: Done

File: docs/idiomatic-maya-guide.md (hunk 2 above)
Comment (sagarkrsd) [reply docs/idiomatic-maya-guide.md]: Done

File: docs/idiomatic-maya-guide.md (hunk 3 above)
Comment (sagarkrsd) [reply docs/idiomatic-maya-guide.md]: Done

File: docs/idiomatic-maya-guide.md (hunk 4 above)
Comment (sagarkrsd) [reply docs/idiomatic-maya-guide.md]: Done

File: docs/idiomatic-maya-guide.md (hunk 5 above)
Comment (sagarkrsd) [reply docs/idiomatic-maya-guide.md]: Done

**Pull request merged**